src/services/item/plugins/invitation/service.ts

```diff
@@ -223,6 +224,34 @@ export class InvitationService {
     return { memberships, invitations };
   }
 
+  /**
+   * Create memberships and invitations given invitation list
+   * @param actor
+   * @param repositories
+   * @param itemId
+   * @param invitations
+   * @returns
+   */
+  async shareItem(
+    actor: Member,
+    repositories: Repositories,
+    itemId: Item['id'],
+    invitations: Pick<Invitation, 'email' | 'permission'>[],
+  ): Promise<{ memberships: ItemMembership[]; invitations: Invitation[] }> {
+    await this.itemService.get(actor, repositories, itemId, PermissionLevel.Admin);
+
+    if (!invitations.length) {
+      throw new NoInvitationReceivedFound();
+    }
+
+    return this._createMembershipsAndInvitationsForUserList(
```

Comment:
Not sure but maybe I would split `_createMembershipsAndInvitationsForUserList` into two functions (createMemberships and createInvitation) and use `_partitionExistingUsersAndNewUsers` directly here.

Reply:
Yes, it's because for me `_createMembershipsAndInvitationsForUserList` as two responsibilities as suggested in its name, it create memberships and invitations. Now if `shareItem` partitions the data and **delegates** the creation of members and invitations to the dedicated functions, it responsibility is only to orchestrate these creations. This could also be done by `_createMembershipsAndInvitationsForUserList` but in my sense this function is useless with these modifications. Otherwise, I have the feeling that `shareItem` is useless, because it does nothing special except calling `_createMembershipsAndInvitationsForUserList` (mostly).

But it's only my point of view 😆

Reply:
I see, I think `_createMembershipsAndInvitationsForUserList ` can be used instead of `service.shareItem`, but the admin access check will then have to happen in the controller, and I'm not really sure of that. So I'll let the implementation like this, we can refactor when we refactor controllers!